docs: add instructions about applying rulesets

tjsilver opened the pull request:
## What does this change?

Adds some instructions on how to apply rulesets for branch protection and status checks for Guardian developers, once they have disabled branch protection when applying this workflow to a repo.

This allows repos to keep the same level of protection as before, but allows for the Scala release app to bypass and run successfully.

Update: the instructions now link out to a page on the guardian/recommendations repo.

adamnfish commented on docs/configuration.md:
```diff
@@ -11,11 +11,13 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
+  on any branch the workflow will be pushing to (ie the default branch). Note that PR #26 means that you can use branch protection rulesets to protect your default branch, so long as you allow your GitHub App to bypass those restrictions.
+### Guardian developers 
+* Set the custom property `production_status` to `production` to apply  branch protection to the default branch via a ruleset (this allows the Scala release app to bypass branch protection).
+* Create a separate 'Status checks' ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging' and add your repo's status check(s) to the 'Status checks that are required' list. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
```
> add your repo's status check(s)

I don't love the term "status checks" here. The status check is the thing being edited, not the thing being referenced. I think this sentence is really referring to their *workflow(s)*?

adamnfish commented on docs/configuration.md:
```diff
@@ -11,11 +11,13 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
+* Set the custom property `production_status` to `production` to apply  branch protection to the default branch via a ruleset (this allows the Scala release app to bypass branch protection).
+* Create a separate 'Status checks' ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging' and add your repo's status check(s) to the 'Status checks that are required' list. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
+![status_checks.png](status_checks.png)
+* Also add the Scala release app to the ruleset bypass list as in the branch protection ruleset.
```
Do we need the word "Also" here?

Is this enough information for someone not super familiar with the topic to find and select the correct app?

NovemberTang commented on docs/configuration.md:
```diff
@@ -11,11 +11,14 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
-  branch protection **rulesets** to protect your default branch, so long as you allow your GitHub App
-  to bypass those restrictions.
-* **Guardian developers:** comply with the repository requirements of
+  on any branch the workflow will be pushing to (ie the default branch). Note that PR #26 means that you can use branch protection rulesets to protect your default branch, so long as you allow your GitHub App to bypass those restrictions.
```
```suggestion
  on any branch the workflow will be pushing to (ie the default branch). Note that PR #26 means that you can use rulesets to protect your default branch, so long as you allow your GitHub App to bypass those restrictions.
```

Also, as this specific library relies on a specific GH App, could we name it explicitly?

NovemberTang commented on docs/configuration.md:
```diff
@@ -11,11 +11,14 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
-* **Guardian developers:** comply with the repository requirements of
+  on any branch the workflow will be pushing to (ie the default branch). Note that PR #26 means that you can use branch protection rulesets to protect your default branch, so long as you allow your GitHub App to bypass those restrictions.
+### Guardian developers 
+* Re-enable branch protections:
```
I think calling this re-enabling is potentially confusing, as we're enabling a slightly different thing.
```suggestion
* Enable branch protection rulesets:
```
might be a bit clearer?

NovemberTang commented on docs/configuration.md:
```diff
@@ -11,11 +11,14 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
+### Guardian developers 
+* Re-enable branch protections:
+  * Set the custom property `production_status` to `production` to apply  branch protection to the default branch via a ruleset (this allows the Scala release app to bypass branch protection).
+  * Create a separate 'Status checks' ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging' and add your repo's workflow name(s) to the 'Status checks that are required' list. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
```
Someone unfamiliar with rulesets will probably assume a 'Status checks' ruleset means something specific, rather than the arbitrary(-ish) name we've assigned to the ruleset.

```suggestion
  * Create a second ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging'. Type your repo's workflow name(s) (not to be confused with the filename) into the 'Status checks that are required' box. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
```
The first sentence is a bit long so maybe we could break it up a bit? I've suggested a couple of other bits that I think would help with clarity

NovemberTang commented on docs/configuration.md:
```diff
@@ -11,11 +11,14 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
+* Re-enable branch protections:
+  * Set the custom property `production_status` to `production` to apply  branch protection to the default branch via a ruleset (this allows the Scala release app to bypass branch protection).
+  * Create a separate 'Status checks' ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging' and add your repo's workflow name(s) to the 'Status checks that are required' list. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
+  ![status_checks.png](status_checks.png)
```
Putting in this diagram was a good call!

NovemberTang commented on docs/configuration.md:
```diff
@@ -11,11 +11,14 @@ The release workflow needs a `release.yml` GitHub workflow in your repo, and spe
+  * Set the custom property `production_status` to `production` to apply  branch protection to the default branch via a ruleset (this allows the Scala release app to bypass branch protection).
+  * Create a separate 'Status checks' ruleset in your repo with the Branch protection property 'Require status checks to pass' -> 'Require branches to be up to date before merging' and add your repo's workflow name(s) to the 'Status checks that are required' list. For example, if your repo has a CI workflow with the name 'CI', then the setting would look like this:
+  ![status_checks.png](status_checks.png)
+  * Add the Scala release app to the ruleset bypass list (see the 'Branch protection' ruleset for an example).
```
Duplicate of an earlier comment, could we reference the app by name either here or above?

tjsilver replied:
I wasn't sure if referencing the app in a public repo was a good idea - so deliberately kept it vague - and Guardian users will see the name of the app in the ruleset created by adding the custom property. WDYT @NovemberTang ?

rtyley commented:
Thanks for this! I do have a few points:

* I would _like_ to make this workflow and it's documentation usable by other organisations, not just the Guardian, so having large amounts of Guardian-specific stuff in the middle of the general Configuration document is quite intrusive. For ease of Guardian developers using the documentation, I'm certainly happy with the occasional single line giving Guardian-specific information, but for more extensive instructions, like these, I think it would be better to link off to a Guardian-specific document elsewhere?
* The general instructions about setting up branch Rulesets (push protection and CI requirement) are likely to apply to many _other_ Guardian repos (Scala apps, Typescript apps & libraries, etc), not just Scala library repos, so they probably don't belong in `gha-scala-library-release-workflow`, but should be well-documented elsewhere! The bypass exemption for the `gu-scala-library-release` bot definitely _does_ need to be documented, and that could live with that main Ruleset documentation, and be referenced with a pointer in the docs in `gha-scala-library-release-workflow`
* I was a bit surprised we're asking devs to manually setup a CI Ruleset on each repo - that feels like something RepoCop might be able to do? I haven't been part of the discussion on RepoCop's role on that, I probably missed something!

rtyley commented:
> Most of the time, this is just `CI`, but this isn't true enough of the time for us to be able to generalise. If they have a second workflow called `test` or `migrate` or anything else, it won't be run, and you'll have to set up a custom ruleset for the repo to catch that anyway. Conversely, if the repo doesn't have a status check called CI - and we've specified that we want it to look for one - it'll hang forever.

Yeah, that makes sense - I guess to an extent standardisation would help here - like, maybe teams should just standardise on having a CI workflow called 'CI'? RepoCop could check for the presence of a 'CI' workflow, and if one exists, set the ruleset up, and if it doesn't, _perhaps_ warn teams that it would be good to standardise on 'CI' - though there may be legitimate cases where complex projects have multiple CI jobs and there is no obvious single test to look at.

tjsilver commented:
>     * I would _like_ to make this workflow and its documentation usable by other organisations, not just the Guardian, so having large amounts of Guardian-specific stuff in the middle of the general Configuration document is quite intrusive. For ease of Guardian developers using the documentation, I'm certainly happy with the occasional single line giving Guardian-specific information, but for more extensive instructions, like these, I think it would be better to link off to a Guardian-specific document elsewhere?
>
>     * The general instructions about setting up branch Rulesets (push protection and CI requirement) are likely to apply to many _other_ Guardian repos (Scala apps, Typescript apps & libraries, etc), not just Scala library repos, so they probably don't belong in `gha-scala-library-release-workflow`, but should be well-documented elsewhere! The bypass exemption for the `gu-scala-library-release` bot definitely _does_ need to be documented, and that could live with that main Ruleset documentation, and be referenced with a pointer in the docs in `gha-scala-library-release-workflow`

@rtyley I have moved the instructions to a new page in the recommendations repo [PR](https://github.com/guardian/recommendations/pull/163), and just added an extra point to the guardian instructions - hope this is ok?

rtyley reviewed (approved):
Yeah, brilliant! Sorry it took me so long to get back to you, this is great 👍
